File: tools/cli/internal/capability/capability.go
// Package capability is the interface-capability-declaration engine from
// issue #128 (§65–§74, §102, §105) and grill Q2d — the representation-plane
// contract by which a surface (web panel, terminal, voice, phone widget)
// declares what it can render, so the system never aims a state at a surface
// that cannot present it. Full model: docs/interface-capabilities.md.
//
// The engine is deliberately pure (the internal/routing pattern): it
// validates declarations, classifies event names, and decides delivery per
// (declaration, event) with a reasoned verdict. No I/O, no clock, no
// transport — the live SSE path applies these decisions at its broadcast
// choke points.
//
// Two invariants:
//
//  1. Subtraction only. A declaration can only narrow what its own
//     connection receives. Decide never returns suppress for an undeclared
//     client, never consults anything but the one declaration handed to it,
//     and grants no send/aim authority — which is why declaring requires no
//     new guarded route.
//  2. Fail loud. An invalid declaration is an error, never a silent
//     fallback to legacy delivery: fail-open would widen what a narrowing
//     surface receives, the exact wrong failure mode (the
//     routing.Policy.Validate posture).
package capability

import (
	"encoding/json"
	"sort"
)

// Class buckets every SSE event name for the delivery gate
// (docs/interface-capabilities.md "The delivery gate").
type Class string

const (
	// ClassLifecycle: connection bootstrap frames. Never gated — a surface
	// that cannot receive `connected` cannot negotiate at all.
	ClassLifecycle Class = "lifecycle"
	// ClassStateReport: the server reporting its own persisted or derived
	// state (message, history, agents, presence_map, …). Never gated in v1:
	// rendering a report it does not care about is a surface's own no-op.
	ClassStateReport Class = "state-report"
	// ClassPresentationCommand: the server aiming an action at a surface.
	// Gated: delivered to a declared surface only if it accepts the name.
	ClassPresentationCommand Class = "presentation-command"
)

// presentationCommands is the gated class — exactly the five bespoke
// panel-aiming names grill Q2d generalizes (07_ARCHITECTURE-GRILL.md).
// Admitting a new name (tts_event and plugin RPC names are the known
// candidates) is a minor schema bump plus an entry here, deliberately not
// done ahead of a declaring consumer (gap-fill 1 in the doc).
var presentationCommands = map[string]bool{
	"navigate":     true,
	"reload":       true,
	"device_cmd":   true,
	"input_action": true,
	"draft":        true,
}

// Classify buckets one SSE event name. Unknown names are state reports by
// construction: only a name listed in the gated class can ever be withheld,
// so a new event is deliverable-to-everyone until deliberately admitted.
func Classify(event string) Class {
	switch {
	case event == "connected":
		return ClassLifecycle
	case presentationCommands[event]:
		return ClassPresentationCommand
	default:
		return ClassStateReport
	}
}

// PresentationCommands returns the gated names, sorted — the roster the
// connected-echo reports as enforceable (Recognize).
func PresentationCommands() []string {
	names := make([]string, 0, len(presentationCommands))
	for n := range presentationCommands {
		names = append(names, n)
	}
	sort.Strings(names)
	return names
}

// Surface is the declaring surface's identity — the vocabulary seam shared
// with the source-contracts work (docs/interface-capabilities.md "Scope
// disambiguation"): kind names the surface family, instance the concrete
// endpoint (the panel passes its ?device= uuid).
type Surface struct {
	Kind     string `json:"kind"`
	Instance string `json:"instance,omitempty"`
}

// Declaration is one surface's capability declaration, exchanged once at
// connect and immutable for the connection's lifetime (LSP posture).
// Unknown top-level JSON fields are ignored on parse — additive minor-bump
// fields must not break an older server.
type Declaration struct {
	// Schema: SemVer of the declaration contract itself
	// (docs/interface-capabilities.md "Versioning").
	Schema  string  `json:"schema"`
	Surface Surface `json:"surface"`
	// Accepts: presentation-command names this surface executes, each with
	// an open detail object (LSP-style nesting) reserved for per-command
	// granularity. The only axis enforced in v1.
	Accepts map[string]json.RawMessage `json:"accepts,omitempty"`
	// Content: content types the surface can present. Advisory in v1 —
	// validated, registered, exposed; not yet gating.
	Content []string `json:"content,omitempty"`
	// Interactions: affordances the surface can hand back. Advisory in v1;
	// load-bearing once interaction-state routing (#128 §64–§66) lands.
	Interactions []string `json:"interactions,omitempty"`
}

// AcceptNames returns the declared accepts, sorted.
func (d *Declaration) AcceptNames() []string {
	names := make([]string, 0, len(d.Accepts))
	for n := range d.Accepts {
		names = append(names, n)
	}
	sort.Strings(names)
	return names
}

// Recognize splits a declaration's accepts into the names this engine
// actually enforces (members of the gated class) and the
// preserved-but-inert rest, both sorted — the payload of the
// connected-echo, so a surface can tell when a capability it cares about
// is not enforced by the server it reached.
func (d *Declaration) Recognize() (recognized, unknown []string) {
	recognized, unknown = []string{}, []string{}
	for _, n := range d.AcceptNames() {
		if presentationCommands[n] {
			recognized = append(recognized, n)
		} else {
			unknown = append(unknown, n)
		}
	}
	return recognized, unknown
}

// Verdict is a delivery gate answer.
type Verdict string

const (
	VerdictDeliver  Verdict = "deliver"
	VerdictSuppress Verdict = "suppress"
)

// Reason names why a verdict came out the way it did — decisions must be
// explainable (#128's observability posture, as in routing.Result).
type Reason string

const (
	// ReasonLegacy: no declaration — grandfathered full delivery.
	ReasonLegacy Reason = "legacy-undeclared"
	// ReasonUngated: the event's class is not gated.
	ReasonUngated Reason = "ungated-class"
	// ReasonAccepted: gated event, declared in accepts.
	ReasonAccepted Reason = "accepted"
	// ReasonNotAccepted: gated event, absent from accepts — the one path
	// that suppresses.
	ReasonNotAccepted Reason = "not-accepted"
)

// Decision is one delivery-gate evaluation.
type Decision struct {
	Verdict Verdict `json:"verdict"`
	Reason  Reason  `json:"reason"`
	Class   Class   `json:"class"`
	Event   string  `json:"event"`
}

// Decide applies the delivery gate for one client and one event. A nil
// declaration is the undeclared legacy client and always delivers; a
// declared client is suppressed exactly when the event is a presentation
// command absent from its accepts (docs/interface-capabilities.md, gate
// table).
func Decide(decl *Declaration, event string) Decision {
	class := Classify(event)
	d := Decision{Verdict: VerdictDeliver, Class: class, Event: event}
	switch {
	case decl == nil:
		d.Reason = ReasonLegacy
	case class != ClassPresentationCommand:
		d.Reason = ReasonUngated
	default:
		if _, ok := decl.Accepts[event]; ok {
			d.Reason = ReasonAccepted
		} else {
			d.Verdict, d.Reason = VerdictSuppress, ReasonNotAccepted
		}
	}
	return d
}

File: packages/server/src/capability.ts
// Interface capability declaration — the OUTPUT-direction delivery gate
// (docs/interface-capabilities.md). A connecting surface may declare, via
// `?caps=<url-encoded JSON>` on GET /api/chat/events, which presentation
// commands it accepts; delivery of the gated event names to that connection
// is then narrowed to its declaration. An undeclared client is legacy:
// byte-identical delivery, gated by nothing.
//
// This is the TS mirror of the normative engine in
// tools/cli/internal/capability (a separate Go module this package cannot
// import). Keep the two in lockstep — the Go package's tests are the
// contract; a divergence here is a bug here.

export const SUPPORTED_SCHEMA_MAJOR = 1
export const MAX_DECLARATION_BYTES  = 8 * 1024
export const MAX_ACCEPTS            = 64
export const MAX_TOKENS             = 32

// The gated class: exactly the five Q2d panel-aiming names. Every other name
// (lifecycle, state reports, unknown/future events) is delivered to everyone —
// a new event is deliverable-to-all until deliberately admitted here.
export const PRESENTATION_COMMANDS = new Set([
  "navigate", "reload", "device_cmd", "input_action", "draft",
])

const NAME_RE     = /^[a-z][a-z0-9_]{0,63}$/
const INSTANCE_RE = /^[A-Za-z0-9._-]{1,128}$/
// Mirrors supersession.ParseVersion's strictness: three plain non-negative
// decimal fields, no signs, no prerelease tags, no leading "v".
const SCHEMA_RE   = /^[0-9]+\.[0-9]+\.[0-9]+$/

export type CapabilityDeclaration = {
  schema:       string
  surface:      { kind: string; instance?: string }
  accepts:      Record<string, unknown>   // name → open detail object, preserved as-is
  content:      string[]
  interactions: string[]
}

// parseDeclaration parses and validates one raw declaration. Strict and loud:
// an invalid declaration must refuse the connection rather than fall back to
// legacy full delivery — fail-open would widen what a narrowing surface
// receives. Unknown top-level fields are deliberately ignored (LSP's posture)
// so an additive field from a newer surface does not break this server.
export function parseDeclaration(raw: string): { decl: CapabilityDeclaration } | { error: string } {
  if (new TextEncoder().encode(raw).length > MAX_DECLARATION_BYTES)
    return { error: `capability declaration exceeds the ${MAX_DECLARATION_BYTES}-byte cap` }
  let parsed: unknown
  try { parsed = JSON.parse(raw) } catch { return { error: "capability declaration is not valid JSON" } }
  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed))
    return { error: "capability declaration: want a JSON object" }
  const obj = parsed as Record<string, unknown>

  const schema = obj.schema
  if (typeof schema !== "string" || !SCHEMA_RE.test(schema))
    return { error: `schema ${JSON.stringify(schema)}: want MAJOR.MINOR.PATCH` }
  const major = parseInt(schema, 10)
  if (major !== SUPPORTED_SCHEMA_MAJOR)
    return { error: `schema major ${major} unsupported (this server implements ${SUPPORTED_SCHEMA_MAJOR})` }

  const surface = obj.surface
  if (typeof surface !== "object" || surface === null || Array.isArray(surface))
    return { error: "surface: want an object with a kind" }
  const kind = (surface as Record<string, unknown>).kind
  if (typeof kind !== "string" || !NAME_RE.test(kind))
    return { error: `surface.kind ${JSON.stringify(kind)}: want ${NAME_RE}` }
  const instance = (surface as Record<string, unknown>).instance
  if (instance !== undefined && instance !== "" && (typeof instance !== "string" || !INSTANCE_RE.test(instance)))
    return { error: `surface.instance ${JSON.stringify(instance)}: want ${INSTANCE_RE}` }

  const rawAccepts = obj.accepts ?? {}
  if (typeof rawAccepts !== "object" || rawAccepts === null || Array.isArray(rawAccepts))
    return { error: "accepts: want an object of name → detail" }
  const accepts = rawAccepts as Record<string, unknown>
  const names = Object.keys(accepts)
  if (names.length > MAX_ACCEPTS)
    return { error: `${names.length} accepts entries exceeds the ${MAX_ACCEPTS} cap` }
  for (const n of names) {
    if (!NAME_RE.test(n)) return { error: `accepts name ${JSON.stringify(n)}: want ${NAME_RE}` }
  }

  const tokens: Record<string, string[]> = {}
  for (const axis of ["content", "interactions"] as const) {
    const arr = obj[axis] ?? []
    if (!Array.isArray(arr)) return { error: `${axis}: want an array of tokens` }
    if (arr.length > MAX_TOKENS)
      return { error: `${arr.length} ${axis} entries exceeds the ${MAX_TOKENS} cap` }
    for (const tok of arr) {
      if (typeof tok !== "string" || !NAME_RE.test(tok))
        return { error: `${axis} token ${JSON.stringify(tok)}: want ${NAME_RE}` }
    }
    tokens[axis] = arr as string[]
  }

  return {
    decl: {
      schema,
      surface: { kind, ...(typeof instance === "string" && instance !== "" ? { instance } : {}) },
      accepts,
      content:      tokens.content,
      interactions: tokens.interactions,
    },
  }
}

// recognize splits a declaration's accepts names into those this server gates
// on and those it has never heard of — echoed back on `connected` so a surface
// can see which of its names are inert here. Unknown names are preserved,
// inert, and reported; never an error (forward compatibility across minors).
export function recognize(decl: CapabilityDeclaration): { recognized: string[]; unknown: string[] } {
  const recognized: string[] = []
  const unknown:    string[] = []
  for (const n of Object.keys(decl.accepts)) {
    (PRESENTATION_COMMANDS.has(n) ? recognized : unknown).push(n)
  }
  recognized.sort()
  unknown.sort()
  return { recognized, unknown }
}

// shouldDeliver is the delivery gate (docs/interface-capabilities.md, gate
// table). The subtraction invariant: a declaration can only remove deliveries
// relative to a legacy client, never add one.
export function shouldDeliver(decl: CapabilityDeclaration | undefined, event: string): boolean {
  if (!decl) return true                              // legacy: grandfathered byte-identical
  if (!PRESENTATION_COMMANDS.has(event)) return true  // lifecycle + state reports are ungated
  return Object.prototype.hasOwnProperty.call(decl.accepts, event)
}

// Suppression counters — a silent gate is indistinguishable from a gate that
// never runs (the presenceBroadcasts precedent). Exposed on
// /api/chat/subscribers as `capability_suppressed`.
const suppressed = new Map<string, number>()

export function countSuppressed(event: string): void {
  suppressed.set(event, (suppressed.get(event) ?? 0) + 1)
}

export function suppressedCounts(): Record<string, number> {
  return Object.fromEntries([...suppressed.entries()].sort(([a], [b]) => a.localeCompare(b)))
}
